File: algorithms/greedy_algorithms/different_summands.py
# python3
'''
The goal of this problem is to represent a given positive integer n as a sum of as many pairwise
distinct positive integers as possible.
'''


def naive_find_n_prizes(n):
    prizes = []
    k = 0

    while n != 0:
        k += 1
        n -= k

        if n < 0:
            n += prizes.pop()

        prizes.append(k)

    return prizes


if __name__ == '__main__':
    n = int(input())
    prizes = naive_find_n_prizes(n)
    print(len(prizes))
    print(*prizes, sep=' ')

File: algorithms/greedy_algorithms/different_summands.cpp
// The goal of this problem is to represent a given positive integer n as a sum of as many pairwise
// distinct positive integers as possible.

#include <iostream>
#include <vector>

std::vector<int64_t> find_numbers(int64_t n) {
    std::vector<int64_t> nums;
    int i = 1;
    
    while (n > 2*i) {
        n -= i;
        nums.push_back(i);
        i++;
    }

    nums.push_back(n);

    return nums;
}

int main() {
    int64_t n;
    std::vector<int64_t> numbers;
    std::cin >> n;
    numbers = find_numbers(n);

    std::cout << numbers.size() << std::endl;
    
    for (int i = 0; i < numbers.size(); i++) {
        std::cout << numbers[i] << " ";
    }

    return 0;
}
